src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,158 @@
+
+            if (!_visible)
+            {
+                return;
+            }
+
+            _closeFromHide = true;
+
+            try
+            {
+                _infoBarUIElement?.Close();
+                _infoBarUIElement = null;
+            }
+            finally
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected async Task<IVsInfoBarHost> GetInfoBarHostAsync(CancellationToken cancellationToken)
+        {
+            var uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+            if (uiShell == null)
+            {
+                return null;
+            }
+
+            // Ensure that we are on the UI thread before interacting with the Solution Explorer UI Element
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+            {
+                return null;
+            }
+
+            object tempObject;
+            if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+            {
+                return null;
+            }
+
+            return tempObject as IVsInfoBarHost;
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
+                new IVsInfoBarTextSpan[] {
+                    new InfoBarTextSpan(Resources.InfoBar_TextMessage),
+                },
+                new IVsInfoBarActionItem[] {
+                    new InfoBarHyperlink(Resources.InfoBar_HyperlinkMessage),
+                },
+                KnownMonikers.StatusWarning);
+        }
+
+        internal async Task ShowInfoBarAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            try
+            {
+                var infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    return;
+                }
+
+                var infoBarHost = await GetInfoBarHostAsync(cancellationToken);
+                if (infoBarHost == null)
+                {
+                    return;
+                }
+
+                // Ensure that we are on the UI thread before interacting with the UI
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+                var infoBarModel = GetInfoBarModel();
+
+                _infoBarUIElement = infoBarFactory.CreateInfoBar(infoBarModel);
```

Comment:
are we null-safe here?

Reply:
Its going to be when I add `#nullable enable`